python/ray/dag/tests/experimental/test_torch_tensor_dag.py

```diff
@@ -292,8 +315,94 @@ def test_torch_tensor_nccl_wrong_shape(ray_start_regular):
 
     compiled_dag.teardown()
 
-    ray.kill(sender)
-    ray.kill(receiver)
+    # TODO(swang): This currently requires time.sleep to avoid some issue with
+    # following tests.
+    time.sleep(3)
+
+
+@pytest.mark.parametrize("ray_start_regular", [{"num_cpus": 4}], indirect=True)
+def test_torch_tensor_nccl_nested(ray_start_regular):
+    """
+    Test nested torch.Tensor passed via NCCL. Its shape and dtype is statically
+    declared.
+    """
+    if not USE_GPU:
+        pytest.skip("NCCL tests require GPUs")
+
+    assert (
+        sum(node["Resources"].get("GPU", 0) for node in ray.nodes()) > 1
+    ), "This test requires at least 2 GPUs"
+
+    actor_cls = TorchTensorWorker.options(num_gpus=1)
+
+    sender = actor_cls.remote()
+    receiver = actor_cls.remote()
+
+    shape = (10,)
+    dtype = torch.float16
+
+    with InputNode() as inp:
+        dag = sender.send_dict_with_tuple_args.bind(inp)
+        dag = dag.with_type_hint(
+            TorchTensorType(shape=shape, dtype=dtype, transport="nccl")
+        )
+        dag = receiver.recv_dict.bind(dag)
+
+    compiled_dag = dag.experimental_compile()
+
+    for i in range(3):
+        args = (shape, dtype, 1)
+
+        output_channel = compiled_dag.execute(args)
+        # TODO(swang): Replace with fake ObjectRef.
+        result = output_channel.begin_read()
+        expected_result = {0: (0, shape, dtype)}
+        assert result == expected_result
+        output_channel.end_read()
+
+    compiled_dag.teardown()
+
+
+@pytest.mark.parametrize("ray_start_regular", [{"num_cpus": 4}], indirect=True)
+def test_torch_tensor_nccl_nested_dynamic(ray_start_regular):
```

Comment:
I am trying to understand the difference between `test_torch_tensor_nccl_dynamic` and `test_torch_tensor_nccl_nested_dynamic`. The only difference between them is `direct_return`. My understanding is:

* In `test_torch_tensor_nccl_dynamic`, 
  * Because `direct_return` is set to True, it doesn't have any `contains_type`.
  * Because `transport` is set to `"nccl"`, it will create a `TorchTensorNcclChannel` instead of a shared memory channel.

* In `test_torch_tensor_nccl_nested_dynamic`,
  * Because `direct_return` is set to False, the DAGNode calls `DAGNode._type_hint.set_contains_type(TorchTensorType)`.
  * [Question] What's the type of `_type_hint`? It seems to be ChannelOutputType ([code](https://sourcegraph.com/github.com/ray-project/ray@d091bb5df7b0a8ec10d11ed53f44129f621c0c7d/-/blob/python/ray/dag/dag_node.py?L66)), and ChannelOutputType's `create_channel` is not implemented ([code](https://sourcegraph.com/github.com/ray-project/ray@d091bb5df7b0a8ec10d11ed53f44129f621c0c7d/-/blob/python/ray/experimental/channel/common.py?L51)).
    * I guess the expected result of this test is that a `NestedTorchTensorNcclChannel` should be created, but I didn't  know how does it create because of the above question.

Reply:
Yes that is all correct.

ChannelOutputType is an interface type and the subclasses are meant to implement `create_channel`. The         self._default_type_hint: ChannelOutputType = SharedMemoryType( for DAG node is a [SharedMemoryType](https://github.com/ray-project/ray/blob/master/python/ray/experimental/channel/shared_memory_channel.py#L81). SharedMemoryType.create_channel creates a `NestedTorchTensorNcclChannel` in this PR if it contains a TorchTensorType.